src/download.c

```diff
@@ -434,23 +438,56 @@ static int perform_curl_io_and_complete(int *left)
  * add new transfer up until the queue reaches the high threshold. At this point
  * we don't return to the caller and instead process the queue until its len
  * gets below the low threshold */
-static int poll_fewer_than(int xfer_queue_high, int xfer_queue_low)
+static int poll_fewer_than(size_t xfer_queue_high, size_t xfer_queue_low)
 {
-	int left;
 	CURLMcode curlm_ret;
+	int running;
 
-	curlm_ret = curl_multi_perform(mcurl, &left);
+	curlm_ret = curl_multi_perform(mcurl, &running);
 	if (curlm_ret != CURLM_OK) {
 		return -1;
 	}
 
-	if (left <= xfer_queue_high) {
+	// There are not enough transfers in queue yet
+	if (mcurl_size <= xfer_queue_high) {
 		return 0;
 	}
 
-	while (left > xfer_queue_low) {
-		usleep(500); /* TODO this really ought to be a select() statement */
-		if (perform_curl_io_and_complete(&left) != 0) {
+	/* When the high bound for the hysteresis is reached, process events
+	 * until the low bound is reached */
+	while (mcurl_size > xfer_queue_low) {
+		int numfds = 0;
+
+		// Wait for activity on the multi-stack
+		curlm_ret = curl_multi_wait(mcurl, NULL, 0, 500, &numfds);
+		if (curlm_ret != CURLM_OK) {
+			return -1;
+		}
+
+		// Either a timeout was hit, or no events to process
+		if (numfds <= 0) {
```

Comment:
Since the last argument to `curl_multi_wait()` is the address of an `int` (signed), I wanted to cover the < 0 case for this reason. The < 0 case makes no sense to me, but I wanted to inspect the curl source to verify it will return >= 0 in all conditions.